cosmology/GravityManager.h

```diff
@@ -0,0 +1,370 @@
+        double relError = std::fabs((M_m - (*rho).sum()) / M_m);
+        mes << "relative error: " << relError << endl;
+
+        size_type TotalParticles = 0;
+        size_type localParticles = this->pcontainer_m->getLocalNum();
+
+        ippl::Comm->reduce(localParticles, TotalParticles, 1, std::plus<size_type>());
+
+        if (ippl::Comm->rank() == 0) {
+            if (TotalParticles != totalP_m || relError > 1e-10) {
+                mes << "Time step: " << it_m << endl;
+                mes << "Total particles in the sim. " << totalP_m << " "
+                    << "after update: " << TotalParticles << endl;
+                mes << "Rel. error in mass conservation: " << relError << endl;
+                ippl::Comm->abort();
+            }
+        }
+
+        // Convert mass assignment to actual mass density
+        double cellVolume = std::reduce(hr.begin(), hr.end(), 1., std::multiplies<double>());
+        (*rho)            = (*rho) / cellVolume;
+        rhoNorm_m         = norm(*rho);
+
+        // rho = rho_e - rho_i (only if periodic BCs)
+        if (this->fsolver_m->getStype() != "OPEN") {
+            double size = 1;
+            for (unsigned d = 0; d < Dim; d++) {
+                size *= rmax[d] - rmin[d];
+            }
+            *rho = *rho - (M_m / size);
+        }
+    }
+
+protected:
+    size_type totalP_m;        ///< Total number of particles.
+    int nt_m;                  ///< Number of time steps.
+    Vector_t<int, Dim> nr_m;   ///< Number of grid points in each dimension.
+    double lbt_m;              ///< Load balance threshold.
+    std::string solver_m;      ///< Solver type.
+    std::string stepMethod_m;  ///< Time stepping method type.
+
+    double time_m;                   ///< Current simulation time.
+    double dt_m;                     ///< Time step size.
+    double a_m;                      ///< Scaling factor.
+    double Dloga;                    ///< Logarithmic increment of the scaling factor.
+    double Hubble_m;                 ///< Hubble constant at the current time [s^-1].
+    double Hubble0;                  ///< Hubble constant today (73.8 km/sec/Mpc).
+    double G;                        ///< Gravitational constant.
+    double rho_crit0;                ///< Critical density today.
+    double O_m;                      ///< Matter density parameter.
+    double O_L;                      ///< Dark energy density parameter.
+    double t_L;                      ///< Characteristic time scale.
+    double z_m;                      ///< Initial redshift.
+    double z_f;                      ///< Final redshift.
+    int it_m;                        ///< Current iteration number.
+    Vector_t<double, Dim> rmin_m;    ///< Minimum comoving coordinates [kpc/h].
+    Vector_t<double, Dim> rmax_m;    ///< Maximum comoving coordinates [kpc/h].
+    Vector_t<double, Dim> hr_m;      ///< Grid spacing in each dimension.
+    double M_m;                      ///< Total mass.
```

Comment:
since units were added for above quantities, would be nice to add them for these too (G, m, t)